gadgets/trace_exec/program.bpf.c

```diff
@@ -148,17 +148,6 @@ int ig_execve_e(struct trace_event_raw_sys_enter *ctx)
 	return 0;
 }
 
-// struct ovl_inode defined in fs/overlayfs/ovl_entry.h
-// Unfortunately, not exported to vmlinux.h
-// and not available in /sys/kernel/btf/vmlinux
-// See https://github.com/cilium/ebpf/pull/1300
-// So we don't benefit from CO-RE's preserve_access_index
-// We only rely on vfs_inode and __upperdentry relative positions
-struct _ovl_inode {
-	struct inode vfs_inode;
-	struct dentry *__upperdentry;
-};
```

Comment:
Perhaps it can be fixed by adding `__attribute__((preserve_access_index));` to this structure? (I think it'll force clang to emit a relocation when the container_of macro needs to calculate `offsetof(type,member)`)

Reply:
Also, why is this structure so different from the upstream one? https://github.com/torvalds/linux/blob/v5.0/fs/overlayfs/ovl_entry.h#L91-L105

Reply:
I think I now understand, we have the inode, and we suppose the dentry is the next member on the structure, so that's why we don't care about any other members.

Reply:
> Perhaps it can be fixed by adding __attribute__((preserve_access_index)); to this structure? (I think it'll force clang to emit a relocation when the container_of macro needs to calculate offsetof(type,member))

cilium/ebpf emits a bad relocation when I add this attribute because it does not find `struct _ovl_inode` in the kernel BTF.
```
const badRelo = 0xbad2310 
```

> I think I now understand, we have the inode, and we suppose the dentry is the next member on the structure, so that's why we don't care about any other members.

Correct. In that way, no need to define types for other fields.

Reply:
> cilium/ebpf emits a bad relocation when I add this attribute because it does not find struct _ovl_inode in the kernel BTF.

That's unfortunate.

Reply:
You can use "flavour names" like `struct ovl_inode___gadget`. Note the three `_`. Anything after the underscore is ignored by the library when trying to find a target.

Reply:
@lmb Thanks for the link, I didn't know about struct flavors.

The following seems to work. Tested without https://github.com/cilium/ebpf/pull/1300, so `struct ovl_inode` is unknown to BTF and only the "else" branch is tested.

Is it the approach you recommend?

```
#include <vmlinux.h>

struct ovl_inode___gadget {
       struct inode vfs_inode;
       struct dentry *__upperdentry;
} __attribute__((preserve_access_index));
...

	if (bpf_core_type_exists(struct ovl_inode___gadget)) {
		struct ovl_inode___gadget *oi = container_of(
			inode, struct ovl_inode___gadget, vfs_inode);
		bpf_probe_read_kernel(&upperdentry, sizeof upperdentry,
				      &oi->__upperdentry);
	} else {
		bpf_probe_read_kernel(&upperdentry, sizeof upperdentry,
				      ((void *)inode) +
					      bpf_core_type_size(struct inode));
	}
```

Reply:
I think something like checking for the existence of the field would be more reliable: https://github.com/libbpf/libbpf/blob/f81eef23b33c0dbf923e863a72ce51ea4d32e291/src/bpf_core_read.h#L151-L164

Reply:
I gave it a try: https://github.com/inspektor-gadget/inspektor-gadget/pull/2406

About `bpf_core_field_exists`: I'll give it a try. If the struct does not exist, I guess `bpf_core_field_exists` would just return false?